File: python/567_PermutationInString.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s1) > len(s2):
            return False
        
        char_list_s1 = [0 for i in range(26)]
        for l in s1:
            char_list_s1[ord(l)-97] += 1
        
        m = len(s1)
        n = len(s2)
        
        curr_char_list = [0 for i in range(26)]
        for l in s2[:m]:
            curr_char_list[ord(l)-97] += 1
        
        for i in range(m, n):
            if curr_char_list == char_list_s1:
                return True
            
            # sliding window
            curr_char_list[ord(s2[i-m])-97] -= 1
            curr_char_list[ord(s2[i])-97] += 1
        
        return curr_char_list == char_list_s1 # covers cases where n == m or permutation occurs s[n-m:]

File: cpp/567_PermutationInString.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.length() > s2.length()) return false;
        
        vector<int> s1_lets(26, 0);
        
        for (char c : s1) ++s1_lets[(int) (c - 'a')];
        
        int m = s1.length();
        int n = s2.length();
        
        vector<int> check_lets(26, 0);
        for (int i = 0; i < m; ++i) ++check_lets[(int) (s2[i] - 'a')];
        
        for (int i = m; i < n; ++i) {
            if (check_lets == s1_lets) return true;
            
            ++check_lets[(int) (s2[i] - 'a')], --check_lets[(int) (s2[i-m] - 'a')];
        }
        return check_lets == s1_lets;
    }
};
